### ISSUE TITLE
How to do supervised contrastive learning using the NTXent loss?

### ISSUE THREAD

**Issue by Waterkin** [KevinMusgrave/pytorch-metric-learning#536]
The official template is like this:

```
from pytorch_metric_learning import miners, losses
import torch
miner = miners.MultiSimilarityMiner()
loss_func = losses.NTXentLoss()

# your training loop
for i, (data, labels) in enumerate(dataloader):
	optimizer.zero_grad()
	logits = model(data)
	hard_pairs = miner(embeddings, labels)
	loss = loss_func(embeddings, labels, hard_pairs)
	loss.backward()
	optimizer.step()
```

How should I sample the `data` here? Should I randomly sample the positive and negative samples or how?
Thanks in advance!

**Comment by KevinMusgrave** [comment]
Random sampling works fine. In other words, using `torch.utils.data.DataLoader(dataset, shuffle=True)` should work, unless your dataset is very imbalanced. For example, if 99% of your data belongs to 1 class, then many batches will contain only 1 class, so no negative pairs will exist in those batches.

Another approach that is sometimes used, is to create class-balanced batches. For example, if the batch size is 64, each batch could contain 8 samples of 8 classes. [MPerClassSampler](https://kevinmusgrave.github.io/pytorch-metric-learning/samplers/#mperclasssampler) does this:

```python
from pytorch_metric_learning.samplers import MPerClassSampler

# labels[i] should return the label for the ith element of your dataset
# m is the number of samples per class, per batch
sampler = MPerClassSampler(labels, m=8, batch_size=64)
dataloader = torch.utils.data.DataLoader(dataset, sampler=sampler)
```

**Comment by Waterkin** [comment]
> Random sampling works fine. In other words, using `torch.utils.data.DataLoader(dataset, shuffle=True)` should work, unless your dataset is very imbalanced. For example, if 99% of your data belongs to 1 class, then many batches will contain only 1 class, so no negative pairs will exist in those batches.
> 
> Another approach that is sometimes used, is to create class-balanced batches. For example, if the batch size is 64, each batch could contain 8 samples of 8 classes. [MPerClassSampler](https://kevinmusgrave.github.io/pytorch-metric-learning/samplers/#mperclasssampler) does this:
> 
> ```python
> from pytorch_metric_learning.samplers import MPerClassSampler
> 
> # labels[i] should return the label for the ith element of your dataset
> # m is the number of samples per class, per batch
> sampler = MPerClassSampler(labels, m=8, batch_size=64)
> dataloader = torch.utils.data.DataLoader(dataset, sampler=sampler)
> ```

Thanks for your reply, Kevin!
